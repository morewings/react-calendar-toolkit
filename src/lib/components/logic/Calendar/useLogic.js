import {useCallback} from 'react';
import {
  ceilMonth,
  checkIsSameMonth,
  checkIsWeekend,
  checkIsWithinInterval,
  floorMonth,
  matchDatesWithPrecision,
} from 'lib/utils/dateUtils';

const useLogic = ({
  disableDate,
  endDate,
  precision,
  startDate,
  highlightDate,
  onDateSet,
  highlightWeekends,
  visibleTimestamp,
  selectedTimestamp,
  todayTimestamp,
}) => {
  const getIsDisabled = useCallback(
    date => {
      if (precision === 'year') return false; // we are not disabling years, just rendering the range
      if (precision === 'month') {
        return !checkIsWithinInterval(
          {start: floorMonth(startDate), end: ceilMonth(endDate)},
          date
        );
      }
      return (
        disableDate({date, isWeekend: checkIsWeekend(date), precision}) ||
        !checkIsWithinInterval({start: startDate, end: endDate}, date)
      );
    },
    [disableDate, endDate, precision, startDate]
  );

  const getIsHighlighted = useCallback(
    date => highlightDate({date, isWeekend: checkIsWeekend(date), precision}),
    [highlightDate, precision]
  );

  const handleDateSet = useCallback(
    date => {
      const isDisabled = getIsDisabled(date);
      !isDisabled && onDateSet(date);
    },
    [getIsDisabled, onDateSet]
  );

  const getIsWeekend = date =>
    precision === 'day' && highlightWeekends && checkIsWeekend(date);

  const getBelongsToSameMonth = date =>
    precision === 'day' && checkIsSameMonth(date, visibleTimestamp);

  const getIsSelected = date =>
    matchDatesWithPrecision(precision, date, selectedTimestamp);

  const getIsCurrent = date =>
    matchDatesWithPrecision(precision, date, todayTimestamp);

  return {
    getIsDisabled,
    getIsHighlighted,
    handleDateSet,
    getIsWeekend,
    getBelongsToSameMonth,
    getIsSelected,
    getIsCurrent,
  };
};

export default useLogic;
