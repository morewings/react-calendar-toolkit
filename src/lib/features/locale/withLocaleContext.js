import React, {useContext} from 'react';
import PropTypes from 'prop-types';
import defaultLocale from 'date-fns/locale/en-US';

const LocaleContext = React.createContext({});

export const useLocaleContext = () => useContext(LocaleContext);

export const Provider = ({children, dateFnsLocale}) => (
  <LocaleContext.Provider value={dateFnsLocale}>
    {children}
  </LocaleContext.Provider>
);

Provider.propTypes = {
  dateFnsLocale: PropTypes.shape({}),
  children: PropTypes.element.isRequired,
};
Provider.defaultProps = {
  dateFnsLocale: defaultLocale,
};

const withLocaleContext = WrappedComponent => ({
  // eslint-disable-next-line react/prop-types
  dateFnsLocale,
  ...restProps
}) => (
  <Provider dateFnsLocale={dateFnsLocale}>
    <WrappedComponent {...restProps} />
  </Provider>
);

export default withLocaleContext;
