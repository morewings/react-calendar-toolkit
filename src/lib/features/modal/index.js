export {initialState} from './ModalReducer';
export {useModalContext} from './context';
export {default as useModalActions} from './actionCreators';
export {default as withModalProvider} from './withModalProvider';
