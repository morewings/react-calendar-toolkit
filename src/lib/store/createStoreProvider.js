import React, {useReducer} from 'react';

const createStoreProvider = ({initialState, reducer, context}) => ({
  // eslint-disable-next-line react/prop-types
  children,
}) => {
  const {Provider} = context;
  const [state, dispatch] = useReducer(reducer, initialState);
  return <Provider value={{state, dispatch}}>{children}</Provider>;
};

export default createStoreProvider;
