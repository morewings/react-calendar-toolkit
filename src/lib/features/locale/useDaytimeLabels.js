import {formatDateWithLocale} from 'lib/utils/dateUtils';
import {useLocaleContext} from './withLocaleContext';

const useDaytimeLabels = () => {
  const locale = useLocaleContext();
  const formatDate = formatDateWithLocale({locale});
  const testString = formatDate('p', new Date(2016, 0, 1, 0));
  return {
    amLabel: formatDate('aaa', new Date(2016, 0, 1, 0)),
    pmLabel: formatDate('aaa', new Date(2016, 0, 1, 16)),
    timeFormat: testString.length > 5 ? '12' : '24',
  };
};

export default useDaytimeLabels;
