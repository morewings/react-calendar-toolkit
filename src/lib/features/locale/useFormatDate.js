import {formatDateWithLocale} from 'lib/utils/dateUtils';
import {useLocaleContext} from './withLocaleContext';

const useFormatDate = () => {
  const locale = useLocaleContext();
  return formatDateWithLocale({locale});
};

export default useFormatDate;
