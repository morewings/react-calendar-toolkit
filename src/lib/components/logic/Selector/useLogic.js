import {useCallback} from 'react';
import {
  ceilMonth,
  checkIsWithinInterval,
  subMonth,
  floorMonth,
  addMonth,
} from 'lib/utils/dateUtils';
import {useDatepickerActions} from 'lib/features/datepicker';

const useLogic = ({startDate, endDate, visibleTimestamp}) => {
  const {setVisibility, setPrecision} = useDatepickerActions();

  const isAvailable = useCallback(
    date =>
      checkIsWithinInterval(
        {start: floorMonth(startDate), end: ceilMonth(endDate)},
        date
      ),
    [endDate, startDate]
  );

  const onSetPrecision = useCallback(
    nextPrecision => {
      setPrecision(nextPrecision);
    },
    [setPrecision]
  );

  const onIncrementMonth = useCallback(() => {
    const nextDate = addMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  const onDecrementMonth = useCallback(() => {
    const nextDate = subMonth(visibleTimestamp, 1);
    isAvailable(nextDate) && setVisibility(nextDate);
  }, [isAvailable, setVisibility, visibleTimestamp]);

  return {onSetPrecision, onIncrementMonth, onDecrementMonth};
};

export default useLogic;
