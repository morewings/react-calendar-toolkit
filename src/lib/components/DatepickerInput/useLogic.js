import {useCallback} from 'react';
import {useModalActions} from 'lib/features/modal';
import {convertToDate} from 'lib/utils/dateUtils';
import {useFormatDate} from 'lib/features/locale';
import {useDatepickerContext} from 'lib/features/datepicker';

const useLogic = ({onDateSet, hideOnSelect}) => {
  const {toggleDatepicker} = useModalActions();
  const formatDate = useFormatDate();
  const {
    state: {selectedTimestamp},
  } = useDatepickerContext();

  const handleDateSet = useCallback(
    date => {
      onDateSet(date);
      hideOnSelect && toggleDatepicker(false);
    },
    [hideOnSelect, onDateSet, toggleDatepicker]
  );

  const formatValue = useCallback(
    formatPattern =>
      formatDate(formatPattern, convertToDate(selectedTimestamp)),
    [formatDate, selectedTimestamp]
  );

  const getDate = () => convertToDate(selectedTimestamp);

  return {handleDateSet, formatValue, getDate};
};

export default useLogic;
