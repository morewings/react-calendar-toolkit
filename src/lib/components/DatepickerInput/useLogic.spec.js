import React from 'react';
import {renderHook, act} from '@testing-library/react-hooks';
import {Provider as ProviderLocale} from 'lib/features/locale/withLocaleContext';
import createStoreProvider from 'lib/store';
import DatepickerContext from 'lib/features/datepicker/context';
import {initialState} from 'lib/features/modal';
import ModalContext from 'lib/features/modal/context';
import useLogic from './useLogic';

const mockReducer = jest.fn();
const onDateSet = jest.fn();

const date = new Date(2012, 11, 12);
const selectedTimestamp = date.getTime();

export const ProviderDatepicker = createStoreProvider({
  initialState: {
    selectedTimestamp,
  },
  reducer: jest.fn(),
  context: DatepickerContext,
});

export const ProviderModal = createStoreProvider({
  initialState,
  reducer: mockReducer,
  context: ModalContext,
});

const Provider = ({children, ...restProps}) => (
  <ProviderLocale {...restProps}>
    <ProviderModal {...restProps}>
      <ProviderDatepicker {...restProps}>{children}</ProviderDatepicker>
    </ProviderModal>
  </ProviderLocale>
);

describe('Datepicker > useLogic > handleDateSet', () => {
  beforeEach(() => {
    mockReducer.mockClear();
    onDateSet.mockClear();
  });

  it('returns functions', () => {
    const {result} = renderHook(() => useLogic({}), {
      wrapper: ({children}) => <Provider>{children}</Provider>,
    });
    expect(result.current).toMatchSnapshot();
  });

  describe('formatValue', () => {
    it('formats selectedTimestamp with provided pattern', () => {
      const {result} = renderHook(() => useLogic({}), {
        wrapper: ({children}) => <Provider>{children}</Provider>,
      });
      expect(result.current.formatValue('MMM do')).toBe('Dec 12th');
    });
  });

  describe('handleDateSet', () => {
    it('it invokes callback with provided date', () => {
      const {result} = renderHook(() => useLogic({onDateSet}), {
        wrapper: ({children}) => <Provider>{children}</Provider>,
      });
      result.current.handleDateSet(date);
      expect(onDateSet.mock.calls[0][0]).toBe(date);
    });

    it('toggles visibility, if hideOnSelect === true', () => {
      const {result} = renderHook(
        () => useLogic({onDateSet, hideOnSelect: true}),
        {
          wrapper: ({children}) => <Provider>{children}</Provider>,
        }
      );
      act(() => result.current.handleDateSet(date));
      expect(mockReducer.mock.calls[0][1]).toMatchSnapshot();
    });

    it('does not toggle visibility, if hideOnSelect === false', () => {
      const {result} = renderHook(
        () => useLogic({onDateSet, hideOnSelect: false}),
        {
          wrapper: ({children}) => <Provider>{children}</Provider>,
        }
      );
      act(() => result.current.handleDateSet(date));
      expect(mockReducer).not.toHaveBeenCalled();
    });
  });
});
