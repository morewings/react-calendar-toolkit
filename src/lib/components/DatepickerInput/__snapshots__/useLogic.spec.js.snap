// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker > useLogic > handleDateSet handleDateSet toggles visibility, if hideOnSelect === true 1`] = `
Object {
  "payload": false,
  "type": "TOGGLE_DATEPICKER",
}
`;

exports[`DatePicker > useLogic > handleDateSet returns functions 1`] = `
Object {
  "formatValue": [Function],
  "getDate": [Function],
  "handleDateSet": [Function],
}
`;
